File: Source/ethnicity.py
"""
Ethnicity-Adjusted Carrier Frequency Engine for Mergenix.

Provides population-specific carrier frequency adjustments based on
gnomAD v4.1 data. Supports Bayesian posterior calculations for
carrier risk given genotype and population background.

Supported populations:
    - African/African American
    - East Asian
    - South Asian
    - European (Non-Finnish)
    - Finnish
    - Latino/Admixed American
    - Ashkenazi Jewish
    - Middle Eastern
    - Global
"""

import json
import math
import os

POPULATIONS: list[str] = [
    "African/African American",
    "East Asian",
    "South Asian",
    "European (Non-Finnish)",
    "Finnish",
    "Latino/Admixed American",
    "Ashkenazi Jewish",
    "Middle Eastern",
    "Global",
]

# Default path for the ethnicity frequency data file
_DATA_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "data")
DEFAULT_ETHNICITY_DATA_PATH = os.path.join(_DATA_DIR, "ethnicity_frequencies.json")


def load_ethnicity_data(path: str | None = None) -> dict:
    """
    Load ethnicity frequency data from JSON file.

    Args:
        path: Path to ethnicity_frequencies.json. Uses default if None.

    Returns:
        Parsed JSON dict with "metadata" and "frequencies" keys.
    """
    if path is None:
        path = DEFAULT_ETHNICITY_DATA_PATH
    with open(path) as f:
        return json.load(f)


def get_population_frequency(
    rsid: str,
    population: str,
    ethnicity_data: dict,
) -> float | None:
    """
    Look up the population-specific carrier frequency for a given rsID.

    Falls back to the "Global" frequency if the requested population
    is not present for that variant.

    Args:
        rsid: SNP identifier (e.g. "rs334").
        population: One of the POPULATIONS values.
        ethnicity_data: Loaded ethnicity frequency data dict.

    Returns:
        Carrier frequency as a float, or None if the rsID is not in the data.
    """
    frequencies = ethnicity_data.get("frequencies", {})
    variant = frequencies.get(rsid)
    if variant is None:
        return None

    freq = variant.get(population)
    if freq is not None:
        return float(freq)

    # Fall back to global
    global_freq = variant.get("Global")
    if global_freq is not None:
        return float(global_freq)

    return None


def adjust_carrier_risk(
    base_risk: float,
    population_frequency: float,
    global_frequency: float,
) -> dict:
    """
    Adjust a base carrier risk using the ratio of population to global frequency.

    The adjustment factor is ``population_frequency / global_frequency``.
    The adjusted risk is clamped to [0.0, 1.0].

    Args:
        base_risk: Original carrier risk (0.0-1.0 scale).
        population_frequency: Carrier frequency in the target population.
        global_frequency: Global carrier frequency.

    Returns:
        Dict with keys:
            - "base_risk": the original risk
            - "adjusted_risk": risk after population adjustment
            - "adjustment_factor": ratio used for the adjustment
            - "population_frequency": the population freq used
            - "global_frequency": the global freq used
    """
    if global_frequency <= 0:
        return {
            "base_risk": base_risk,
            "adjusted_risk": base_risk,
            "adjustment_factor": 1.0,
            "population_frequency": population_frequency,
            "global_frequency": global_frequency,
        }

    adjustment_factor = population_frequency / global_frequency
    adjusted_risk = min(max(base_risk * adjustment_factor, 0.0), 1.0)

    return {
        "base_risk": base_risk,
        "adjusted_risk": adjusted_risk,
        "adjustment_factor": adjustment_factor,
        "population_frequency": population_frequency,
        "global_frequency": global_frequency,
    }


def calculate_bayesian_posterior(
    prior_frequency: float,
    genotype_data: dict,
    population: str,
) -> float:
    """
    Compute Bayesian posterior probability of being a carrier given
    genotype observation and population background.

    P(carrier | genotype, population) =
        P(genotype | carrier) * P(carrier) /
        [P(genotype | carrier) * P(carrier) + P(genotype | non-carrier) * P(non-carrier)]

    Genotype likelihoods:
        - "carrier" genotype  -> P(obs | carrier) = 0.99, P(obs | non-carrier) = prior * 0.01
        - "normal"  genotype  -> P(obs | carrier) = 0.01, P(obs | non-carrier) = 0.99
        - "affected" genotype -> P(obs | carrier) = 0.50, P(obs | non-carrier) = 0.001
        - otherwise           -> uniform (returns prior unchanged)

    Args:
        prior_frequency: Population carrier frequency (0-1).
        genotype_data: Dict with at least a "status" key
                       ("carrier", "normal", "affected", or "unknown").
        population: Population label (used for documentation; the prior
                    already encodes population information).

    Returns:
        Posterior probability of being a carrier (0.0-1.0).
    """
    if prior_frequency is None or prior_frequency < 0:
        return 0.0
    if prior_frequency > 1:
        prior_frequency = 1.0

    status = genotype_data.get("status", "unknown") if genotype_data else "unknown"

    # Likelihood P(genotype | carrier) and P(genotype | non-carrier)
    if status == "carrier":
        p_genotype_given_carrier = 0.99
        p_genotype_given_noncarrier = max(prior_frequency * 0.01, 1e-10)
    elif status == "normal":
        p_genotype_given_carrier = 0.01
        p_genotype_given_noncarrier = 0.99
    elif status == "affected":
        p_genotype_given_carrier = 0.50
        p_genotype_given_noncarrier = 0.001
    else:
        # Unknown — no evidence, posterior equals prior
        return prior_frequency

    prior_carrier = prior_frequency
    prior_noncarrier = 1.0 - prior_frequency

    numerator = p_genotype_given_carrier * prior_carrier
    denominator = numerator + p_genotype_given_noncarrier * prior_noncarrier

    if denominator <= 0:
        return 0.0

    posterior = numerator / denominator
    return min(max(posterior, 0.0), 1.0)


def get_ethnicity_summary(
    rsid: str,
    all_populations: list[str] | None,
    ethnicity_data: dict,
) -> dict:
    """
    Build a frequency comparison across all populations for a given variant.

    Args:
        rsid: SNP identifier.
        all_populations: List of population names. Defaults to POPULATIONS.
        ethnicity_data: Loaded ethnicity frequency data dict.

    Returns:
        Dict with keys:
            - "rsid": the queried rsID
            - "gene": gene symbol (if present)
            - "condition": condition name (if present)
            - "frequencies": dict mapping population -> frequency (or None)
            - "global": global frequency (or None)
            - "found": bool indicating if the rsID was in the data
    """
    if all_populations is None:
        all_populations = POPULATIONS

    frequencies_db = ethnicity_data.get("frequencies", {})
    variant = frequencies_db.get(rsid)

    if variant is None:
        return {
            "rsid": rsid,
            "gene": None,
            "condition": None,
            "frequencies": {pop: None for pop in all_populations},
            "global": None,
            "found": False,
        }

    pop_freqs = {}
    for pop in all_populations:
        pop_freqs[pop] = variant.get(pop)

    return {
        "rsid": rsid,
        "gene": variant.get("gene"),
        "condition": variant.get("condition"),
        "frequencies": pop_freqs,
        "global": variant.get("Global"),
        "found": True,
    }


def format_frequency_comparison(
    population_freq: float,
    global_freq: float,
) -> str:
    """
    Format a human-readable comparison of population vs global frequency.

    Examples:
        - "2.3x higher than global average"
        - "0.4x lower than global average"
        - "Equal to global average"

    Args:
        population_freq: Carrier frequency in target population.
        global_freq: Global carrier frequency.

    Returns:
        Formatted comparison string.
    """
    if global_freq is None or global_freq <= 0:
        if population_freq and population_freq > 0:
            return "No global data for comparison"
        return "No data available"

    if population_freq is None or population_freq < 0:
        return "No population data available"

    ratio = population_freq / global_freq

    if math.isclose(ratio, 1.0, rel_tol=0.05):
        return "Equal to global average"
    elif ratio > 1.0:
        return f"{ratio:.1f}x higher than global average"
    else:
        return f"{ratio:.1f}x lower than global average"

File: packages/genetics-engine/src/ethnicity.ts
/**
 * Ethnicity-Adjusted Carrier Frequency Engine
 *
 * Provides population-specific carrier frequency adjustments based on
 * gnomAD v4.1 data. Supports Bayesian posterior calculations for carrier
 * risk given genotype and population background.
 *
 * Supported populations (from gnomAD v4.1):
 * - African/African American
 * - East Asian
 * - South Asian
 * - European (Non-Finnish)
 * - Finnish
 * - Latino/Admixed American
 * - Ashkenazi Jewish
 * - Middle Eastern
 * - Global
 *
 * Ported from Source/ethnicity.py (288 lines).
 */

import type {
  Population,
  EthnicityAdjustment,
  EthnicitySummary,
  CarrierStatus,
  EthnicityFrequenciesData,
} from './types';

import { clamp } from './utils';
import { POPULATIONS } from '@mergenix/genetics-data';

// ─── Population Frequency Lookup ────────────────────────────────────────────

/**
 * Look up the population-specific carrier frequency for a given rsID.
 *
 * Falls back to the "Global" frequency if the requested population is not
 * present for that variant.
 *
 * Ported from Source/ethnicity.py `get_population_frequency()`.
 *
 * @param rsid - SNP identifier (e.g., "rs334")
 * @param population - Target population label (one of POPULATIONS)
 * @param ethnicityData - Loaded ethnicity frequency data
 * @returns Carrier frequency as a number, or null if rsID not in data
 */
export function getPopulationFrequency(
  rsid: string,
  population: Population,
  ethnicityData: EthnicityFrequenciesData,
): number | null {
  const variant = ethnicityData.frequencies[rsid];
  if (variant === undefined) {
    return null;
  }

  // Look up the requested population frequency
  const freq = variant[population] as number | undefined;
  if (freq !== undefined) {
    return freq;
  }

  // Fall back to Global frequency
  const globalFreq = variant.Global as number | undefined;
  if (globalFreq !== undefined) {
    return globalFreq;
  }

  return null;
}

// ─── Risk Adjustment ────────────────────────────────────────────────────────

/**
 * Adjust a base carrier risk using the ratio of population to global frequency.
 *
 * The adjustment factor is (population_frequency / global_frequency).
 * The adjusted risk is clamped to [0.0, 1.0].
 *
 * Ported from Source/ethnicity.py `adjust_carrier_risk()`.
 *
 * @param baseRisk - Original carrier risk (0.0-1.0 scale)
 * @param populationFrequency - Carrier frequency in the target population
 * @param globalFrequency - Global carrier frequency
 * @returns Adjustment result with base risk, adjusted risk, and factor
 */
export function adjustCarrierRisk(
  baseRisk: number,
  populationFrequency: number,
  globalFrequency: number,
): EthnicityAdjustment {
  if (globalFrequency <= 0) {
    return {
      baseRisk,
      adjustedRisk: baseRisk,
      adjustmentFactor: 1.0,
      populationFrequency,
      globalFrequency,
    };
  }

  const adjustmentFactor = populationFrequency / globalFrequency;
  const adjustedRisk = clamp(baseRisk * adjustmentFactor, 0.0, 1.0);

  return {
    baseRisk,
    adjustedRisk,
    adjustmentFactor,
    populationFrequency,
    globalFrequency,
  };
}

// ─── Bayesian Posterior Calculation ─────────────────────────────────────────

/**
 * Compute Bayesian posterior probability of being a carrier given genotype
 * observation and population background.
 *
 * Uses Bayes' theorem:
 *   P(carrier | genotype, population) =
 *     P(genotype | carrier) * P(carrier) /
 *     [P(genotype | carrier) * P(carrier) + P(genotype | non-carrier) * P(non-carrier)]
 *
 * Special cases:
 * - "affected": Returns 1.0 immediately. Affected individuals (homozygous
 *   pathogenic) have two copies and will always pass at least one to offspring.
 *   Computing P(carrier|affected) via Bayes is scientifically meaningless.
 * - "unknown":  Returns prior unchanged (no evidence)
 *
 * Genotype likelihoods (for carrier/normal only):
 * - "carrier" genotype:  P(obs|carrier) = 0.99, P(obs|non-carrier) = prior * 0.01
 * - "normal" genotype:   P(obs|carrier) = 0.01, P(obs|non-carrier) = 0.99
 *
 * Ported from Source/ethnicity.py `calculate_bayesian_posterior()`.
 *
 * @param priorFrequency - Population carrier frequency (0-1)
 * @param genotypeStatus - Carrier status from genotyping ("carrier", "normal", "affected", "unknown")
 * @param _population - Population label (for documentation; prior encodes population)
 * @returns Posterior probability of being a carrier (0.0-1.0)
 */
export function calculateBayesianPosterior(
  priorFrequency: number,
  genotypeStatus: CarrierStatus,
  _population: Population,
): number {
  // Clamp prior: return 0 if negative, cap at 1
  if (priorFrequency == null || priorFrequency < 0) {
    return 0.0;
  }
  let prior = priorFrequency;
  if (prior > 1) {
    prior = 1.0;
  }

  // Unknown status: no evidence, posterior equals prior
  if (genotypeStatus === 'unknown') {
    return prior;
  }

  // Affected individuals (homozygous pathogenic) are definitionally beyond
  // carrier status — they have two copies. Return 1.0 as they will always
  // pass at least one pathogenic allele to offspring.
  if (genotypeStatus === 'affected') {
    return 1.0;
  }

  // Set likelihoods P(genotype | carrier) and P(genotype | non-carrier)
  let pGenotypeGivenCarrier: number;
  let pGenotypeGivenNoncarrier: number;

  if (genotypeStatus === 'carrier') {
    pGenotypeGivenCarrier = 0.99;
    pGenotypeGivenNoncarrier = 0.005;
  } else {
    // "normal": genotype observation suggests non-carrier
    pGenotypeGivenCarrier = 0.01;
    pGenotypeGivenNoncarrier = 0.99;
  }

  const priorCarrier = prior;
  const priorNoncarrier = 1.0 - prior;

  const numerator = pGenotypeGivenCarrier * priorCarrier;
  const denominator = numerator + pGenotypeGivenNoncarrier * priorNoncarrier;

  if (denominator <= 0) {
    return 0.0;
  }

  const posterior = numerator / denominator;
  return clamp(posterior, 0.0, 1.0);
}

// ─── Ethnicity Summary ──────────────────────────────────────────────────────

/**
 * Build a frequency comparison across all populations for a given variant.
 *
 * Ported from Source/ethnicity.py `get_ethnicity_summary()`.
 *
 * @param rsid - SNP identifier
 * @param ethnicityData - Loaded ethnicity frequency data
 * @param populations - List of population labels (defaults to all POPULATIONS)
 * @returns Summary with per-population frequencies
 */
export function getEthnicitySummary(
  rsid: string,
  ethnicityData: EthnicityFrequenciesData,
  populations?: Population[],
): EthnicitySummary {
  const pops: Population[] = populations ?? (POPULATIONS as Population[]);

  const variant = ethnicityData.frequencies[rsid];

  if (variant === undefined) {
    // Build frequencies map with null for every population
    const nullFrequencies = {} as Record<Population, number | null>;
    for (const pop of pops) {
      (nullFrequencies as Record<string, number | null>)[pop] = null;
    }
    return {
      rsid,
      gene: null,
      condition: null,
      frequencies: nullFrequencies,
      global: null,
      found: false,
    };
  }

  // Build per-population frequency map
  const frequencies = {} as Record<Population, number | null>;
  const variantRecord = variant as unknown as Record<string, unknown>;
  for (const pop of pops) {
    const freq = variantRecord[pop];
    (frequencies as Record<string, number | null>)[pop] = typeof freq === 'number' ? freq : null;
  }

  return {
    rsid,
    gene: variant.gene ?? null,
    condition: variant.condition ?? null,
    frequencies,
    global: variant.Global ?? null,
    found: true,
  };
}

/**
 * Format a human-readable comparison of population vs global frequency.
 *
 * Examples:
 * - "2.3x higher than global average"
 * - "0.4x lower than global average"
 * - "Equal to global average"
 *
 * Ported from Source/ethnicity.py `format_frequency_comparison()`.
 *
 * @param populationFreq - Carrier frequency in target population
 * @param globalFreq - Global carrier frequency
 * @returns Formatted comparison string
 */
export function formatFrequencyComparison(
  populationFreq: number | null,
  globalFreq: number | null,
): string {
  // Handle null/zero global frequency
  if (globalFreq == null || globalFreq <= 0) {
    if (populationFreq != null && populationFreq > 0) {
      return 'No global data for comparison';
    }
    return 'No data available';
  }

  // Handle null/negative population frequency
  if (populationFreq == null || populationFreq < 0) {
    return 'No population data available';
  }

  const ratio = populationFreq / globalFreq;

  // Check if approximately equal (within 5% relative tolerance)
  if (Math.abs(ratio - 1.0) / 1.0 <= 0.05) {
    return 'Equal to global average';
  } else if (ratio > 1.0) {
    return `${ratio.toFixed(1)}x higher than global average`;
  } else {
    return `${ratio.toFixed(1)}x lower than global average`;
  }
}
